manager/orchestrator/task.go

```diff
@@ -93,6 +93,32 @@ func IsTaskDirty(s *api.Service, t *api.Task) bool {
 		(t.Endpoint != nil && !reflect.DeepEqual(s.Spec.Endpoint, t.Endpoint.Spec))
 }
 
+// IsTaskPlacementConstraintsOnlyDirty checks if the Placement field alone
+// in the spec has changed.
+func IsTaskPlacementConstraintsOnlyDirty(s *api.Service, t *api.Task) bool {
+	if s.Spec.Task.Placement == nil && t.Spec.Placement == nil {
+		return false
+	}
+
+	// Make a deep copy of the service and task spec for the comparison.
+	serviceTaskSpec := *s.Spec.Task.Copy()
+
+	// Compare the task placement constraints.
+	var placementConstraintsChanged bool
+	if (s.Spec.Task.Placement != nil && t.Spec.Placement == nil) ||
+		(s.Spec.Task.Placement == nil && t.Spec.Placement != nil) ||
+		!reflect.DeepEqual(serviceTaskSpec.Placement, t.Spec.Placement) {
+		placementConstraintsChanged = true
+	}
+
+	// Update spec placement to only the fields
+	// other than the placement constraints in the spec.
+	serviceTaskSpec.Placement = t.Spec.Placement
+	specChanged := !reflect.DeepEqual(serviceTaskSpec, t.Spec)
```

Comment:
Similarly, there are other order-insensitive fields in the service spec (like network attachments) that you need to be careful of here.

Reply:
The current thinking is to blindly compare the rest of the spec and consider it dirty if anything changed. There's definitely other parts of the spec that don't require for the task to be rescheduled. We can enhance this follow up PRs. LMK if you have any concerns with that.